src/server/bitops_family.cc

```diff
@@ -577,12 +593,488 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
   HandleOpValueResult(res, cntx);
 }
 
+enum class EncodingType { UINT, INT, NILL };
+
+struct CommonAttributes {
+  EncodingType type;
+  size_t encoding_bit_size;
+  size_t offset;
+};
+
+struct Nill {};
+using ResultType = std::variant<int64_t, Nill>;
```

Comment:
why is it needed?

Reply:
to use visit instead of if/else. But in many cases it seems like if/else is simpler 😅

Reply:
I prefer using `if/else` than using `visit` @kostasrim

Reply:
We need the special value `Nill`, to denote that one of the subcommands (GET|SET|INCRBY) failed because of overflow/underflow. In that case, we still need to return the value of an empty array. So for example:

```BITFIELD foo SET i1 0 1 OVERFLOW FAIL INCRBY i1 0 1``` 
Will return `{0, null}`. We can't use an `int` or `0` to represent overflow failures and therefore I introduced this Nill value for commands that failed the check. 

I can use an `if/else`, it won't make much of a difference IMO.

```
  1015     auto nill_case = [&](Nill nill) { (*cntx)->SendNull(); };                                                                   
  1016     auto int_cases = [&](auto num) { (*cntx)->SendLong(num); };                                                                 
  1017                                                                                                                                 
  1018     std::visit(Overloaded{nill_case, int_cases}, elem);  
```

vs

```
if(std::holds_alternative<Nill>(elem)) {
    (*cntx)->SendNull();
}
(*cntx)->SendLong(num);
```

But yeah, I will make the change since you find it more helpful

Reply:
The differences are:

1. std::variant is more direct than optional. I either have a `Nill` or a `int64_t`.
2. We can extend in the future by adding overloads without needing to add extra if's. (for now we are feature complete, but we could have new subcommands in the future like MULT, etc which would be far better expressed by a variant)

`Nill` is not `std::monostate` even if they look alike. The first represents a result of a computation. The later is used to allow a std::variant default constructable when the underline types are not default constructor and it HAS to be the first argument in the type list. For example:
```
struct non_default {
   non_default(int) {}
};

std::variant<non_default> var; //compile error
std::variant<non_default, std::monostate> var; //compile error, the default constructor always used the first type argument
std::variant<std::monostate, non_default> var; //Good compiles
```
long story short, I don't want to hijack monostate :stuck_out_tongue: 

But, I am happy to substitute to an if/else and if in the future we need to add more, we will make the simple change :)

Reply:
std::monostate is _literally_ just an empty struct, they're are the same, just different names

> std::variant is more direct than optional. I either have a Nill or a int64_t
  
What's the justification for optional then 🙂 You either carry information or you don't...

Either way, just answering

Reply:
> std::monostate is literally just an empty struct, they're are the same, just different names

Yes, all I am trying to say is that I was trying to be explicit about the intent. `std::monostate` does not mean `Nill` if you don't know the code. The way I think about it is that if I have a monostate it means that I just want to default construct a variant. It doesn't have any other meaning. On the other hand when I use Nill, it does have a meaning, it's a `Nill` result -- allowing the variant to default construct is a nice added property but not the point of Nill since once I use monostate + one type it's basically an optional :stuck_out_tongue: 

I am not arguing here, I am trying to explain how my thinking around it.

p.s. I removed the variant, it's now an optional :)